### ISSUE TITLE
[MU4 Issue] Shortcuts for Adding Sharps/Flats in Input Mode

### ISSUE THREAD

**Issue by jieshunwang** [musescore/MuseScore#11500]
**Describe the bug**
While I tested inputting a score completely from scratch, I really missed having shortcuts which added sharps/flats in input mode. It is a big hassle to have to go back and enter those flats manully using a mouse, especially when the program automatically reverts those accidentals later in the bar. (i.e. if I had two As in a bar, if I attach an accidental to the first A, a natural sign will be automatically applied to the second A)

Since I use a numpad when entering those notes, the top row of the numpad (/ * -) can be great shortcuts for sharps/flats. I am not familiar with all the shortcuts that are available in the program, so I don't know if they are already used up by other functions, but it just seems the most logical place for it to be.

**Comment by Tantacrul** [comment]
@bkunda - would be good to make a recommendation here. It strikes me as really unusual that we don't have primary keyboard shortcuts for accidentals... probably the shortcuts I personally use more than any other!

Once you've had time to think about it, just make a recommendation here and we'll assign it.

**Comment by MarcSabatella** [comment]
FWIW, the usual recommended way of adding accidentals during note input is via the cursor keys - enter G, hit Up to raise it a half step to G#, etc.  That's still the way experienced users probably do it most of the time, and this still works.  So most of the time, you don't need a mouse.  But this method has some significant disadvantages, such as the inability to get courtesy accidentals this way.  So it's good indeed to have shortcuts for adding accidentals directly.

There *are* commands for this; they just don't have shortcuts defined by default, mostly because no one could agree on a good set that seemed both intuitive and reasonable across a variety of keyboard layouts.  In MU3, you could define shortcuts yourself in Edit / Preferences / Shortcuts.  As far as I know, the commands probably exist in MU4 as well, but they don't appear in the dialog because they lack default shortcuts.  This is the bug described in https://github.com/musescore/MuseScore/issues/11060.  So I believe once that bug is fixed, this issue will partially go away.  But, we should still consider assigning shortcuts by default as well.

That's where it gets tricky, though, and to understand the issues, there is some backstory that is good to know.  Short version: there are actually *two* sets of accidental commands to which you can assign shortcuts: one that adds the accidentals "preemptively" and the other "retroactively".

For most of MuseScore's existence, the accidental buttons on the toolbar - and the commands you could assign shortcuts to - worked retroactively in note input mode.  That is, if you wanted G#, first you entered the G, then clicked the sharp button.  But some users were not happy about this, as it can be distracting to hear the G playback first.  And it's objectively inconsistent with the way most of the rest of the note input commands work (eg, you select duration and voice before entering the note, why not accidentals).  So *logically*, it seemed to make sense to change the behavior of the accidental buttons to be "preemptive", and this was done around 3.3 or so.

However, logical as the new behavior may seem, that change proved to be an *incredibly* unpopular decision.  People accustomed to the previous behavior found it very hard to adjust to needing to pre-select the accidental, even after a year or more.  To the point where I think that it is a contributing factor in why some Linux distribution maintainers have refused to update their repositories and are still stuck on 3.2.3 three years later.

After getting beat up long enough from the constant backlash over this, we eventually added the old retroactive behavior back, as separate commands.  So there is now a "preemptive flat" (which is what the toolbar does but can also be assigned a custom shortcut) and a "retroactive flat" command (available by custom shortcut only).  And no, the actual names of the commands don't use those terms; they are even more confusing.

Anyhow, bottom line: when we think about what default shortcuts to use, we also need to consider whether we do this for both sets of commands, and if not, which set to prioritize (although presumably, it would be the "logical" preemptive set).

**Comment by bkunda** [comment]
> So there is now a "preemptive flat" (which is what the toolbar does but can also be assigned a custom shortcut) and a "retroactive flat" command (available by custom shortcut only). And no, the actual names of the commands don't use the terms "preemptive" and "retroactive"; the actual terms used are even more confusing.

@MarcSabatella can you please advise specifically which shortcuts currently apply the retroactive command? The actual terms must indeed be sufficiently confusing that I can't even discover them for myself!

**Comment by MarcSabatella** [comment]
The standard/preemptive versions are labeled "Note input: (flat)"; the retroactive ones are "Note input (non-toggle): (flat)".  In both cases, it's actually the flat sign instead of the word flat, because searching isn't important but saving a translation string is ;-).

Again, neither are actually available in MU4, but that's presumably only because of https://github.com/musescore/MuseScore/issues/11060

**Comment by bkunda** [comment]
OK got it. I now FINALLY understand what those "non-toggle" actions do. 

So it seems we have (in MS3) everything we need to implement shortcuts for both preemptive and retroactive accidentals. I can also see how this could be a really useful thing to have. 

The first step would be for me to chase-up those commands to ensure they get re-instated in MS4.

The second step is to then decide on some sensible shortcuts to provide as defaults. This is where things might get controversial. My initial impulse is to suggest that, whatever shortcuts we designate for these commands "out of the box", that they be useful to users on smaller keyboards as well as those with numpads (Or, in other words, that they don't _just_ work on full-size keyboards with numpads). My second impulse would be to suggest that they be discoverable and sensible to users of non-US/UK keyboard layouts too. 

(On this last point, I'd suggest we steer somewhat clear of Dorico's solution, which is (on the face of it, very logically) to use the `0`, `-`, and `+` keys for natural, flat and sharp accidentals, respectively; these are quite difficult to use on AZERTY and QWERTZ keyboards, not least because these keys aren't always located within proximity of each other. On the other hand, Sibelius's solution of using `7`, `8` and `9` (obviously chosen to map onto their keypad UI component) does, I think, work well on most keyboards. Something in this vein might be the best path for us as well (noting we'd probably need to avoid `7`, which is a current priority mapping for the whole note duration).

The third step would be to rename those "non-toggle" actions so that people can actually understand what they do.

I'm happy to give all this some further thought and make a recommendation. With the above in mind, I'd also be keen for suggestions on what might be appropriate shortcuts to use for these accidentals. I'm thinking that it could be nice for the shortcut keys to be used for both preemptive and retroactive application, but with a modifier key applied for the latter.

**Comment by Tantacrul** [comment]
Do there need to be modifiers to distinguish between preemptive and retroactive?

I mean - we have one of each symbol in the top bar. If you hit it in advance, it turns Note Input on and prepares to add a sharp. The buttons also just work if you select a note and add a shortcut too.

For this reason, if Sharp was (say) mapped to the key: **s**, then you could press **s** in both of these cases without a modifier.

Perhaps I'm misunderstanding the distinction.

**Comment by bkunda** [comment]
> The buttons also just work if you select a note and add a shortcut afterwards too.

The difference with the retroactive application is that you don't need to select the note to apply the shortcut; instead, the shortcut is applied to the note you've just entered, but the note input cursor remains poised to enter a new note. 

This is actually pretty handy if you've entered a note but forgotten to add the accidental. Instead of having to leave input mode to select the note and apply the accidental, you can just quickly tack on the accidental and continue engraving.

**Comment by Tantacrul** [comment]
Ah. I was thinking the answer seemed to simple to be true.

**Comment by MarcSabatella** [comment]
Yes, the problem of defining suitable default shortcuts is messy.  One other set I thought about leveraged the fact that you can enter accidentals within text using Ctrl+Shift+B, Ctrl+Shift+#, and Ctrl+Shift+H.  Of course, Shift+# is just # on most keyboards, but never mind that for now :-).  This would have worked for note input mode, but Ctrl+Shift+B inserts a "B".  And I can't say anyone was ever in love with those shortcuts for text either, but they were convenient because they also allow for consistency with the shortcuts for dynamics characters.

I do kind of like the *idea* of - and + for flat and sharp, but, on my keyboard at least, one requires Shift, the other doesn't, which seems odd.  Plus "+" is well-established as the shortcut for tie.  And "0" is rest.

So you end up with this whole chain reaction of changes to other shortcuts needed in order to implement anything that seems at all logical.  That's why no one has wanted to touch this.

This could actually be a rare case where a shortcut *sequence* might make sense for our default shortcuts.  So, one key as the enabler, then a second to actually select the accidental.  Like, "\ B" or "\ -" to execute the preemptive flat command.  This has the advantage of also allowing easy assignment of similar shortcuts for the retroactive commands - "Shift+\ B" or "Shift+\ -" to execute the retroactive flat command.  But it's really tough to love this idea.

FWIW, my own custom shortcuts in 3.6.2 involve [, ], and \ as flat, sharp, and natural.  I use [ and ] for other commands also, so the accidentals are those with Ctrl+Shift to active the preemptive commands, or in a sequence preceded by tick (" ` ") to active retroactive commands,  I use tick as a sequence enabler for a number of other commands, like "` 1 2" to exchange voices 1 & 2.

**Comment by MarcSabatella** [comment]
TL;DR - please consider my "apply input state" command idea as per https://github.com/musescore/MuseScore/pull/5505.  Which, in the language of this thread, might be called "apply preemptive duration/accidental state retroactively".

**Comment by jieshunwang** [comment]
I always feel awkward doing this, and I hope I'm not intruding too much into the work you guys are doing, but I do hope that I'm contributing somewhat to the improvement of musescore.

I have to say that I have used Sibelius for quite a while and have gotten very used to their shortcuts and that might contribute to my bias in this case. But I do think to make a straightforward and logical shortcut in this case requires giving up somewhat on some other areas.

I see it a good trade off in removing the shortcut of 7 for the whole note duration in favour of utilising it for accidentals instead, because I would assume that a lot more people would benefit from having a shortcut for accidentals than the people who would be inconvenienced by not having a shortcut for the whole note duration. And the fact also is that in the default workspace, the button for the whole note duration doesn't even show up in the top bar.

**Comment by bkunda** [comment]
@jieshunwang I believe the whole note duration _does_ appear by default in the note input bar (at least, it does for me! Perhaps I'm missing something here?)

Thank you for your suggestion about the Sibelius commands – it's actually also something I've thought about too. And in some situations I think it can make a lot of sense to reproduce familiar interactions across different pieces of related software (particularly if the interactions are well-designed). In this case, however, I think the mapping in MuseScore between the number row and the durations in the note input toolbar is really well-established, and it would be difficult to replace the whole note with anything else that doesn't feel completely unrelated to the pattern that's currently there. 

@MarcSabatella I do recall our discussions about your suggestion for an "apply retroactive" command. Did you ever suggest which shortcut key might trigger this? I'd like to give it a test-run 🙂

**Comment by MarcSabatella** [comment]
I doubt it's possible to test the PR directly anymore, someone who knows more about how the current codebase is structured would need to adapt it.  The actual code to do the work is quite simple and would probably still work - it's more the hooking into the command structure that I assume has changed.

As for a default shortcut to use, my PR didn't define one, but what I'd propose is we define a pair of related shortcuts for the existing "apply current palette element" (incredibly useful command with no current default shortcut) and this new one, since they are somehow kind of similar.  For me, the backslash feels like a natural choice for no logical reason all except it's available, so maybe "\" for one, "Shift+\" for the other.  Or single & double quote.  Or semicolon and colon.

**Comment by bkunda** [comment]
> what I'd propose is we define a pair of related shortcuts for the existing "apply current palette element" (incredibly useful command with no current default shortcut) and this new one, since they are somehow kind of similar. For me, the backslash feels like a natural choice for no logical reason all except it's available, so maybe "" for one, "Shift+" for the other. Or single & double quote. Or semicolon and colon.

I'm sorry, maybe it's because this issue discussion is now getting fairly dense but I don't understand this suggestion.

In my simplistic fantasy, could there not just be one simple shortcut (or a combination of two keys) that the user enters when they're in note input mode to apply to the previously entered note whatever duration/accidental is active in the note input toolbar? Something like `Shift`+`R`? ("R" is already associated with the notion of repeating something, which is kind of the idea here). 

This still also doesn't solve our problem for primary shortcuts for accidentals. I'm going to go out on a limb and make the following suggestions:

- `#` = add sharp (Noting that "Shift" will be required on some keyboards) 
- `-` = add flat
- `=` = add natural (Replacing the current assignation `Toggle automatic placement for selected elements`, which I suspect will be far less used than applying an accidental)

These keys will be readily available both on small keyboards and those with numpads. The disadvantage is that the keys may not be located within physical proximity to one another, but I'm hopeful that the visual association of visual elements with how the accidentals work might be intuitive enough to compensate for this.  

I'd like to rule-out adopting shortcuts that cause a significant change to the logic of what we already have in place for frequently-used commands (e.g. changing note duration).

Thoughts?

**Comment by MarcSabatella** [comment]
Yes, I am proposing a single shortcut for the proposed new "apply current input state" command.  I am just observing that the specific shortcut we choose should be one that allows us to also choose a similar one for the *already-existing* "apply current palette element" command, which does not currently have a default shortcut but I believe should and that the two shortcuts should be related.  Just thinking ahead to make sure we don't paint ourselves into a corner with our choice here.

So for instance, if we chose Shift+R for "apply current input state", what does this leave for "apply current palette element"?  Ctrl+R is already taken for "reset" (which is a bit unfortunate because it's somehow almost the opposite of "R").  Ctrl+Shift+R is available I think, but am not crazy about increasing the similarity to reset.  And I gather there are issues with using Alt in default shortcuts, as it is often reserved on Linux systems (although I could be wrong about this).

My think turns next to the word "apply", there are issues with A, P, and L in that we already have shortcuts using these letters that don't seem related at all.  But, hmm, somehow Y works for me (eg, Y for apply current input state, Shift+Y for apply current palette element).  Or as I said, punctuation keys.

On the other hand, we could probably use Ctrl+Shift+F9 for "apply current palette element" to leverage the similarity to other palette commands instead of other "apply" commands.

Regarding the accidental commands themselves, I think -/+/= is as good as anything, and indeed, I don't think there is any great loss in reassigning "toggle autoplace".

@Tantacrul - I recognize some aspects of this discussion are of the sort you are advising against.  I've commented in the forum regarding ideas for how/where these sorts of necessary technical discussions relating to an issue should take place.
